instrumentation/runtime-telemetry-jfr/library/src/main/java/io/opentelemetry/instrumentation/runtimetelemetryjfr/JfrTelemetry.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.runtimetelemetryjfr;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.instrumentation.runtimetelemetryjfr.internal.RecordedEventHandler;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Predicate;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jdk.jfr.EventSettings;
+import jdk.jfr.consumer.RecordingStream;
+
+/** The entry point class for the JFR-over-OpenTelemetry support. */
+public final class JfrTelemetry implements Closeable {
+
+  private static final Logger logger = Logger.getLogger(JfrTelemetry.class.getName());
+
+  private final AtomicBoolean isClosed = new AtomicBoolean();
+  private final OpenTelemetry openTelemetry;
+  private final List<RecordedEventHandler> recordedEventHandlers;
+  private final RecordingStream recordingStream;
+
+  @SuppressWarnings("CatchingUnchecked")
+  JfrTelemetry(OpenTelemetry openTelemetry, Predicate<JfrFeature> featurePredicate) {
+    this.openTelemetry = openTelemetry;
+    this.recordedEventHandlers = HandlerRegistry.getHandlers(openTelemetry, featurePredicate);
+    try {
+      logger.log(Level.INFO, "Starting JfrTelemetry");
+      recordingStream = new RecordingStream();
+      recordedEventHandlers.forEach(
```

Comment:
Yesterday at SIG you asked why `AgentLoadedIntoBootstrapTest` `Agent loads in when separate jvm is launched` test fails with your changes. It currently does not fail anymore because you have disabled the jfr metrics by default. As far as I can tell it fails because `RecordingStream` starts a non-daemon thread. When I used reflection to make it a daemon thread that test passed. I used
```java
Field f = RecordingStream.class.getDeclaredField("directoryStream");
f.setAccessible(true);
Object eventStream = f.get(recordingStream);
Method method = eventStream.getClass().getMethod("setDaemon", boolean.class);
method.invoke(eventStream, true);
```

Reply:
Oh ok I see. The test was failing because the test JVM couldn't shutdown (instead it timed-out) with the background JFR streaming thread running in user mode.

I changed things so that the [JFR streaming thread is a daemon](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7886/files#diff-26c141b781c262d9626416644201b1f889d3a5fffa60c8b13419b96b600fd693R44), but decided not to use reflection.

I was also forgetting to do a proper clean up after the application is finished. I've now [added a shutdown hook here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/7886/files#diff-2b5c8af976a39e6774a9ae67ad64f1309768464c192e9f571f7ee150692a7bacR38) to take care of this.

Thank you!

Reply:
Thanks.